### PR TITLE
DummyLogger can be called with unknown methods

### LINKED ISSUE #13203: DummyLogger AttributeError when calling WandbLogger methods with fast_dev_run
## 🐛 Bug
When using the `fast_dev_run` flag `pytorch-lightning` replaces the logger with a `DummyLogger` object. But with a WandbLogger I also log media with `self.logger.log_text` in my pl module.  I therefore get an AttributeError that the DummyLogger does not have the attribute `log_text`. 

### To Reproduce

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.loggers import WandbLogger


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        self.logger.log_text(
            "samples",
            columns=["epoch", "temperatue", "text"],
            data=[[1, 0.7, "generated_text"]],
        )
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    wandb_logger = WandbLogger(project="myproject", offline=False)

    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        logger=wandb_logger,
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        fast_dev_run=True
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.test(model, dataloaders=test_data)


if __name__ == "__main__":
    run()

```

### Expected behavior

The `DummyLogger` ignores attribute access of external loggers. 

### Environment

```bash

* CUDA:
	- GPU:
	- available:         False
	- version:           None
* Packages:
	- numpy:             1.22.3
	- pyTorch_debug:     False
	- pyTorch_version:   1.11.0
	- pytorch-lightning: 1.6.3
	- tqdm:              4.64.0
* System:
	- OS:                Darwin
	- architecture:
		- 64bit
		-
	- processor:         i386
	- python:            3.9.10
	- version:           Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64
```

### Additional context

Requires probably a `wandb` setup.

cc @borda @awaelchli @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy

### PR DESCRIPTION
**Pull request by HallerPatrick**
## What does this PR do?

This PR overrides the `__getattr__` method of the `DummyLogger` class, such that an unkown method calls due to external 
loggers do not cause an `AttributeError` but return a dummy method and print information about the call to a non-existing method.

Fixes #13203 

### Does your PR introduce any breaking changes? If yes, please list them.

-

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loggers/logger.py [hunk 1]
```diff
@@ -346,6 +346,16 @@ def __iter__(self):
         # if DummyLogger is substituting a logger collection, pretend it is empty
         yield from ()
 
+    def __getattr__(self, name):
+        """Allows the DummyLogger to be called with arbitrary methods, to avoid AttributeErrors."""
+
+        def method(*args, **kwargs):
+            rank_zero_info(
+                f"The DummyLogger is being used and the method `{name}` was called on it but it does not exist"
```
Comment (carmocca) [inline pytorch_lightning/loggers/logger.py:354]: Let's mention fast_dev_run
```suggestion
                f"The `DummyLogger` is being used and the method `{name}` was called on it but it does not exist."
                " This is the case when `fast_dev_run` is enabled"
```

File: pytorch_lightning/loggers/logger.py [hunk 2]
```diff
@@ -346,6 +346,16 @@ def __iter__(self):
         # if DummyLogger is substituting a logger collection, pretend it is empty
         yield from ()
```
Comment (carmocca) [inline pytorch_lightning/loggers/logger.py:347]: Can you update the CHANGELOG? The "Added" section would be the best fit

File: pytorch_lightning/loggers/logger.py [hunk 3]
```diff
@@ -346,6 +346,16 @@ def __iter__(self):
...
 
+    def __getattr__(self, name):
+        """Allows the DummyLogger to be called with arbitrary methods, to avoid AttributeErrors."""
+
+        def method(*args, **kwargs):
+            rank_zero_info(
+                f"The DummyLogger is being used and the method `{name}` was called on it but it does not exist"
+            )
+
+        return method
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/loggers/logger.py:358]: TIL :)

File: pytorch_lightning/loggers/logger.py [hunk 4]
```diff
@@ -346,6 +346,17 @@ def __iter__(self):
...
         yield from ()
 
+    def __getattr__(self, name: str) -> Callable:
+        """Allows the DummyLogger to be called with arbitrary methods, to avoid AttributeErrors."""
+
+        def method(*args, **kwargs):
+            rank_zero_info(
+                f"The `DummyLogger` is being used and the method `{name}` was called on it but it does not exist."
+                " This is the case when `fast_dev_run` is enabled"
+            )
```
Comment (awaelchli) [inline pytorch_lightning/loggers/logger.py:356]: What is the motivation behind emitting an info here? This would lead to a spam of messages for every logging step.
Also it is not only used for fast dev run. Can we drop it? A dummy method with return value None should suffice I think.

File: pytorch_lightning/loggers/logger.py [hunk 5]
```diff
@@ -346,6 +346,17 @@ def __iter__(self):
         # if DummyLogger is substituting a logger collection, pretend it is empty
         yield from ()
 
+    def __getattr__(self, name: str) -> Callable:
```
Comment (awaelchli) [inline pytorch_lightning/loggers/logger.py:349]: Note, now that is is implemented, getattr would be called on any attribute lookup that cant be relsolved by getattribute(). We assume the attribute is always a method, which is the best we can do I guess.

File: pytorch_lightning/loggers/logger.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/loggers/logger.py:356]: I suggested this, because otherwise this could be silently being executed without users noticing

Comment (awaelchli) [reply pytorch_lightning/loggers/logger.py:356]: I think we shouldn't do this. The whole point is that using the dummy logger silently makes all logging calls dumb and do nothing to simulate logging being turned off. If we need a message, it should be done through the trainer when fast dev run is active, IMO, we can say something along the lines of logging and checkpointing is suppressed for fast dev run.

Comment (HallerPatrick) [reply pytorch_lightning/loggers/logger.py]: I rewrote it

Comment (awaelchli) [reply pytorch_lightning/loggers/logger.py]: This looks great! Just the test needs to be adjusted

File: tests/tests_pytorch/loggers/test_logger.py [hunk 6]
```diff
@@ -257,6 +257,18 @@ def test_dummylogger_noop_method_calls():
     logger.log_metrics("1", 2, three="three")
 
 
+def test_dummlogger_arbitrary_method_calls(caplog):
+    """Test that the DummyLogger can be called with non existing methods."""
+    logger = DummyLogger()
+    # Example method from WandbLogger
+    logger.log_text("1", columns=["1", "2", "3"], data=[[1, 2, 3]])
```
Comment (awaelchli) [inline tests/tests_pytorch/loggers/test_logger.py:264]: Since getattr was implemented, you could now assert

```py
assert hasattr(logger, "cocofruit")
assert callable(logger.cocofruit)
```

**Review by awaelchli — APPROVED** [review]
Just the test needs an update

File: CHANGELOG.md [hunk 7]
```diff
@@ -74,6 +74,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 -
 
 
+
+- Override `__getattr__` in `DummyLogger` to avoid `AttributeError` from third party logger calls in `fast_dev_run` ([#13224](https://github.com/PyTorchLightning/pytorch-lightning/pull/13224))
```
Comment (rohitgr7) [inline CHANGELOG.md:78]:
```suggestion
- Added support for calling unknown methods with `DummyLogger` ([#13224](https://github.com/PyTorchLightning/pytorch-lightning/pull/13224))
```

File: src/pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -589,7 +589,10 @@ def _init_debugging_flags(
             self.check_val_every_n_epoch = 1
             self.loggers = [DummyLogger()] if self.loggers else []
 
-            rank_zero_info(f"Running in fast_dev_run mode: will run the requested loop using {num_batches} batch(es).")
+            rank_zero_info(
+                f"Running in fast_dev_run mode: will run the requested loop using {num_batches} batch(es). "
+                "Logging and checkpointing is suppressed."
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:594]:
```suggestion
                f"Running in `fast_dev_run` mode: will run the requested loop using {num_batches} batch(es). "
                "Logging and checkpointing is suppressed."
```

File: tests/tests_pytorch/loggers/test_logger.py [hunk 9]
```diff
@@ -257,6 +257,18 @@ def test_dummylogger_noop_method_calls():
...
+
+    assert (
+        "The `DummyLogger` is being used and the method `log_text` was called on it "
+        "but it does not exist. This is the case when `fast_dev_run` is enabled\n" in caplog.text
+    )
```
Comment (rohitgr7) [inline tests/tests_pytorch/loggers/test_logger.py:269]:
```suggestion
    with caplog.at_level(level=logging.INFO):
        logger.log_text("1")

    assert "method `log_text` was called on it but it does not exist" in caplog.text
```
do import `logging`

File: src/pytorch_lightning/loggers/logger.py [hunk 10]
```diff
@@ -346,6 +346,14 @@ def __iter__(self):
         # if DummyLogger is substituting a logger collection, pretend it is empty
         yield from ()
 
+    def __getattr__(self, name: str) -> Callable:
+        """Allows the DummyLogger to be called with arbitrary methods, to avoid AttributeErrors."""
+
+        def method(*args, **kwargs):
```
Comment (rohitgr7) [inline src/pytorch_lightning/loggers/logger.py:352]: where is the info generated?

File: tests/tests_pytorch/loggers/test_logger.py (hunk 9 above)
Comment (HallerPatrick) [reply tests/tests_pytorch/loggers/test_logger.py:269]: Capturing logging is not required as the dummy method is not calling `rank_zero_info` anymore

File: src/pytorch_lightning/loggers/logger.py (hunk 10 above)
Comment (HallerPatrick) [reply src/pytorch_lightning/loggers/logger.py:352]: Now only general info is generated when running in `fast_dev_run`, see [here](https://github.com/HallerPatrick/pytorch-lightning/blob/2833cc19e2f59f793be9cb91c42a79b6b8f51caf/src/pytorch_lightning/trainer/trainer.py#L594)

File: tests/tests_pytorch/loggers/test_logger.py (hunk 6 above)
Comment (HallerPatrick) [reply tests/tests_pytorch/loggers/test_logger.py:264]: Yes, that looks more clean!

**Review by otaj — APPROVED** [review]
